src/tests/nativeaot/SmokeTests/PInvoke/PInvokeNative.cpp

```diff
@@ -649,6 +649,31 @@ DLL_EXPORT void* __stdcall GetFunctionPointer()
     return (void*)&SetLastErrorFunc;
 }
 
+#ifdef TARGET_WINDOWS
+class IComInterface: public IUnknown
+{
+public:
+    virtual HRESULT DoWork(int param) = 0;
+};
+GUID IID_IComInterface = { 0x111e91ef, 0x1887, 0x4afd, { 0x81, 0xe3, 0x70, 0xcf, 0x08, 0xe7, 0x15, 0xd8 } };
+
+IComInterface* capturedComObject;
+DLL_EXPORT int __stdcall CaptureComPointer(IComInterface* pUnk)
+{
+    if (SUCCEEDED(pUnk->QueryInterface(IID_IComInterface, (void **)&capturedComObject)))
```

Comment:
It the passed in interface is IComInterface already, it would sufficient to just do AddRef here.

Or is there a bug on the managed in that it does not actually pass in IComInterface?

Reply:
My understanding, that CCW always produce IUnknown. At least when I trying to find when QI to desired interface performed by runtime, I cannot. Indirect confirmation of that for me was https://github.com/dotnet/runtimelab/blob/feature/NativeAOT/src/tests/Interop/PInvoke/NativeCallManagedComVisible/Default/DefaultTest.cs#L605
I will change signature of C# and native methods to receive object and IUnknown respectively.

If you know existing code which should pass actual IComInterface and not IUnknown I would be glad to look at.

Reply:
> CCW always produce IUnknown.

That's true for `[MarshalAs(UnmanagedType.IUnknown)] object o)`.

If the signature uses interface, the object should be marshalled as the desired interface. A real world example is here: https://github.com/dotnet/runtime/blob/79ae74f5ca5c8a6fe3a48935e85bd7374959c570/src/libraries/Common/src/Interop/Windows/Ole32/Interop.GetHGlobalFromStream.cs#L13

So `CaptureComPointer(IComInterface foo)` should be marshalled as `IComInterface`, not as `IUnknown`.

Reply:
Then this is my fault. sigh. Maybe I can implement just `[MarshalAs(UnmanagedType.IUnknown)] object o)` scenario then, otherwise I can swamp in the Marshallers world. This seems to be separate scenario for internal implementation, so I anyway will go deeper.

Regarding changes which I should do to make `IComInterface`. I will have to change `InteropHelpers.ConvertManagedComInterfaceToNative(object pUnk)` to `InteropHelpers.ConvertManagedComInterfaceToNative(object pUnk, Guid guid)` probably, or in worst case `InteropHelpers.ConvertManagedComInterfaceToNative(object pUnk, Type targetType)`.
Latter one require me to call `Type.GUID` which probable require reflection.

Reply:
Yes, that's correct.